pkg/operator/ceph/csi/spec.go

```diff
@@ -662,3 +681,198 @@ func (r *ReconcileCSI) validateCSIVersion(ownerInfo *k8sutil.OwnerInfo) (*CephCS
+	// This appears to be the only supported configuration. (See call to applyCephClusterNetworkConfig)
+	cephClusters, err := r.context.RookClientset.CephV1().CephClusters(r.opConfig.OperatorNamespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrap(err, "failed to find CephClusters")
+	}
+	for _, cephCluster := range cephClusters.Items {
+		// Assuming one CephCluster in the r.opConfig.OperatorNamespace
+		if cephCluster.Spec.Network.IsMultus() {
+			if publicNetwork, ok = cephCluster.Spec.Network.Selectors["public"]; !ok {
+				logger.Debug("Public network not provided. Not performing multus configuration.")
+				return nil
+			}
+			isMultusApplied = true
+			break
+		}
+	}
+
+	if !isMultusApplied {
+		logger.Debug("Multus not used. No additional network configuration necassary.")
+		return nil
+	}
+
+	// Populate the host network namespace with a multus-connected interface.
+	tp := templateParam{
+		Namespace: r.opConfig.OperatorNamespace,
+	}
+
+	tp.MultusNetwork = publicNetwork
+	tp.MultusImage = r.opConfig.Image
+
+	multusHostSetup, err := templateToDaemonSet(csiMultusSetup, MultusDaemonsetTemplatePath, tp)
+	if err != nil {
+		return errors.Wrap(err, "failed to get daemonset template to set up the csi network")
+	}
+
+	// Applying affinity and toleration to multus daemonset.
+	// The multus network must be configured on every node that csi plugin pods run on.
+	pluginTolerations := getToleration(r.opConfig.Parameters, pluginTolerationsEnv, []corev1.Toleration{})
+	pluginNodeAffinity := getNodeAffinity(r.opConfig.Parameters, pluginNodeAffinityEnv, &corev1.NodeAffinity{})
+	applyToPodSpec(&multusHostSetup.Spec.Template.Spec, pluginNodeAffinity, pluginTolerations)
+
+	err = k8sutil.CreateDaemonSet(csiMultusSetup, r.opConfig.OperatorNamespace, r.context.Clientset, multusHostSetup)
+	if err != nil {
+		return errors.Wrap(err, "failed to start daemonset to set up the csi network")
+	}
+
+	// Wait for daemonset pods to come up.
+	// Will poll daemonset to ensure pods are available for interface migration.
+	err = wait.Poll(5*time.Second, 30*time.Second, func() (bool, error) {
+		daemonset, err := r.context.Clientset.AppsV1().DaemonSets(r.opConfig.OperatorNamespace).Get(ctx, "csi-multus", metav1.GetOptions{})
+		if err != nil {
+			return false, errors.Wrap(err, "failed to get multus daemonset")
+		}
+
+		if daemonset.Status.NumberReady == daemonset.Status.DesiredNumberScheduled {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Deploy jobs to move network interface from daemonset pods to host network r.opConfig.OperatorNamespace.
+	podList, err := r.context.Clientset.CoreV1().Pods(r.opConfig.OperatorNamespace).List(ctx, metav1.ListOptions{
+		LabelSelector: "app=rook-ceph-multus",
+	})
+	if err != nil || len(podList.Items) < 1 {
+		return errors.Wrap(err, "failed to get multus daemonset pods")
+	}
+
+	var migrationErrorStr string
+	for _, pod := range podList.Items {
```

Comment:
This looks a bit slow waiting for each job before moving to the next one. I believe some concurrency is needed. How about running a goroutine for each with a WaitGroup and then wait for completions?

Reply:
Agreed, this will be impactful if there's a lot of nodes.